firmware/brake/kia_soul_ev/src/brake_control.cpp

```diff
@@ -0,0 +1,196 @@
+ *
+ */
+#define SENSOR_VALIDITY_CHECK_FAULT_COUNT ( 4 )
+
+
+static void read_brake_pedal_position_sensor(
+    brake_pedal_position_s * const value );
+
+
+void check_for_operator_override( void )
+{
+    if ( g_brake_control_state.enabled == true
+        || g_brake_control_state.operator_override == true )
+    {
+        brake_pedal_position_s brake_pedal_position;
+
+        read_brake_pedal_position_sensor( &brake_pedal_position );
+
+        uint32_t brake_pedal_position_average =
+            (brake_pedal_position.low + brake_pedal_position.high) / 2;
+
+        if ( brake_pedal_position_average >= BRAKE_PEDAL_OVERRIDE_THRESHOLD )
+        {
+            disable_control( );
+
+            publish_fault_report( );
+
+            g_brake_control_state.operator_override = true;
+
+            DEBUG_PRINTLN( "Operator override" );
+        }
+        else
+        {
+            g_brake_control_state.operator_override = false;
+        }
+    }
+}
+
+
+void check_for_sensor_faults( void )
+{
+    if ( (g_brake_control_state.enabled == true)
+        || DTC_CHECK(g_brake_control_state.dtcs, OSCC_BRAKE_DTC_INVALID_SENSOR_VAL) )
+    {
+        static int fault_count = 0;
+
+        brake_pedal_position_s brake_pedal_position;
+
+        read_brake_pedal_position_sensor( &brake_pedal_position );
+
+        // sensor pins tied to ground - a value of zero indicates disconnection
+        if( (brake_pedal_position.high == 0)
+            || (brake_pedal_position.low == 0) )
+        {
+            ++fault_count;
+
+            if( fault_count >= SENSOR_VALIDITY_CHECK_FAULT_COUNT )
+            {
+                disable_control( );
+
+                publish_fault_report( );
+
+                DTC_SET(
+                    g_brake_control_state.dtcs,
+                    OSCC_BRAKE_DTC_INVALID_SENSOR_VAL );
+
+                DEBUG_PRINTLN( "Bad value read from brake pedal position sensor" );
+            }
+        }
+        else
+        {
+            DTC_CLEAR(
+                    g_brake_control_state.dtcs,
+                    OSCC_BRAKE_DTC_INVALID_SENSOR_VAL );
+
+            fault_count = 0;
+        }
+    }
+}
+
+
+void update_brake(
+    uint16_t spoof_command_high,
+    uint16_t spoof_command_low )
+{
+    if ( g_brake_control_state.enabled == true )
+    {
+        uint16_t spoof_high =
+            constrain(
+                spoof_command_high,
+                BRAKE_SPOOF_HIGH_SIGNAL_RANGE_MIN,
+                BRAKE_SPOOF_HIGH_SIGNAL_RANGE_MAX );
+
+        uint16_t spoof_low =
+            constrain(
+                spoof_command_low,
+                BRAKE_SPOOF_LOW_SIGNAL_RANGE_MIN,
+                BRAKE_SPOOF_LOW_SIGNAL_RANGE_MAX );
+
+        if( (spoof_high > BRAKE_LIGHT_SPOOF_HIGH_THRESHOLD)
+            || (spoof_low > BRAKE_LIGHT_SPOOF_LOW_THRESHOLD) )
+        {
+            cli();
+            digitalWrite(PIN_BRAKE_LIGHT_ENABLE, HIGH);
+            sei();
+        }
+        else
+        {
+            cli();
```

Comment:
Is there any reason not to write your own atomic digital read/write routines? Could help reduce the number of explicit interrupt enable/disables you have to manage.

Reply:
The only issue I see is the overhead involved with calling a function, although they could be inline functions to help with that. However, since other hardware-interacting pieces also have `cli()` and `sei()`, I like the consistency of keeping the digital/analog read/write calls in `cli()` and `sei()` blocks as well, just to make it explicit.

Reply:
Works for me. On your next refactor opportunity, you might consider taking advantage of the avr-libc convenience macros for doing atomically executed code blocks (http://www.nongnu.org/avr-libc/user-manual/group__util__atomic.html).